### PR TITLE
*: refactor clientv3 balancer, upgrade gRPC to v1.7.2

### PR DESCRIPTION
**Pull request by gyuho**
If we want to notify only healthy addresses, we need a way to look up an endpoint is healthy or not, which is tracked in `healthBalancer`. This combines `healthBalancer` into `simpleBalancer`. Plus, they already have many redundant fields (e.g. `eps`, `hostPort2eps`, etc.).

Fix https://github.com/coreos/etcd/issues/8828.

### CODE REVIEW

**Comment by gyuho** [comment]
gRPC is still broken (master, v1.7.x) from tests in https://github.com/coreos/etcd/pull/8841.

**Comment by xiang90** [comment]
let us focus on existing test. not #8841

**Comment by xiang90** [comment]
can we remove the new test out of this?

**Comment by gyuho** [comment]
Without https://github.com/coreos/etcd/pull/8841, still fails

```
--- FAIL: TestKVGetResetLoneEndpoint (10.70s)
	kv_test.go:891: context deadline exceeded
	kv_test.go:900: timed out waiting for Get
	testing.go:699: race detected during execution of test
```

Basically, `TestBalancerUnderServerStopInflightLinearizableGetOnRestartPinLeaderStopPinLater ` in https://github.com/coreos/etcd/pull/8841 was replacing `TestKVGetResetLoneEndpoint`.

Seems like gRPC sub-connection state change handler has some problem with transient failure.

Will investigate in a separate branch.

**Comment by xiang90** [comment]
@gyuho why this is closed?

**Comment by gyuho** [comment]
@xiang90 Was rebasing from master. Will force push, after running CIs in my VMs.

**Comment by gyuho** [comment]
I ran integration test suites several times in other VMs (all pass), other than coverage tests...

https://github.com/coreos/etcd/blob/75c7e62dc7ce9a4cc2c7ace8f596e8384d779d36/test#L189-L194

```
=== RUN   TestCtlV2Backup
--- FAIL: TestCtlV2Backup (2.64s)
	ctl_v2_test.go:290: unexpected output (got lines [], line count 13)
```

Do we want to block on this?

**Comment by xiang90** [comment]
No. The coverage thing is a separate issue I believe.

I am going to review the whole thing tomorrow, and hopefully get this merged.

File: clientv3/retry.go [hunk 1]
```diff
@@ -62,6 +62,11 @@ func (c *Client) newRetryWrapper(isStop retryStopErrFunc) retryRPCFunc {
 			if logger.V(4) {
 				logger.Infof("clientv3/retry: error %q on pinned endpoint %q", err.Error(), pinned)
 			}
+			// retry when initial connection has not been established
+			// grpc/grpc-go >v1.7.x (balancer v1 wrapper)
+			if rpctypes.ErrorDesc(err) == "there is no connection available" {
```
Comment (xiang90) [inline clientv3/retry.go]: make the error msg as a const?

File: clientv3/health_balancer_test.go [hunk 2]
```diff
@@ -66,8 +66,8 @@ func TestBalancerGetUnblocking(t *testing.T) {
 	}
 
 	down1(errors.New("error"))
-	if addrs := <-sb.Notify(); len(addrs) != len(endpoints) {
-		t.Errorf("closing the only connection should triggered balancer to send the all endpoints via Notify chan so that we can establish a connection")
+	if addrs := <-sb.Notify(); len(addrs) != 2 {
```
Comment (xiang90) [inline clientv3/health_balancer_test.go]: len(addrs) != len(endpoints) - 1 // we call down on one endpoint

File: clientv3/health_balancer_test.go [hunk 3]
```diff
@@ -121,8 +121,8 @@ func TestBalancerGetBlocking(t *testing.T) {
...
```
Comment (xiang90) [inline clientv3/health_balancer_test.go]: len(addrs) != len(endpoints) - 1 // we call down on one endpoint

**Comment by gyuho** [comment]
TODO: will add back if tests still fail. Then this also need to be added to stop check on mutable ops.

```diff
diff --git a/clientv3/retry.go b/clientv3/retry.go
index e6d17d032..699191ff3 100644
--- a/clientv3/retry.go
+++ b/clientv3/retry.go
@@ -62,6 +62,11 @@ func (c *Client) newRetryWrapper(isStop retryStopErrFunc) retryRPCFunc {
                        if logger.V(4) {
                                logger.Infof("clientv3/retry: error %q on pinned endpoint %q", err.Error(), pinned)
                        }
+                       // retry when initial connection has not been established
+                       // grpc/grpc-go >v1.7.x (balancer v1 wrapper)
+                       if rpctypes.ErrorDesc(err) == "there is no connection available" {
+                               continue
+                       }

                        if s, ok := status.FromError(err); ok && (s.Code() == codes.Unavailable || s.Code() == codes.DeadlineExceeded || s.Code() == codes.Internal) {
                                // mark this before endpoint switch is triggered
```

**Comment by xiang90** [comment]
lgtm if CI is reliably passing.

**Comment by gyuho** [comment]
I run full test suites (same as semaphore) in 4 other VMs. And all passed. Merging.

**Comment by xarses** [comment]
I'm wondering if the gRPC proxy uses this enough to put in front of older clients so that I can get this behavior until I can get them on updated to use this directly.

**Comment by gyuho** [comment]
Copying from slack conversation:

gRPC Proxy is implemented as an etcd client wrapper, so if the gRPC Proxy is built with latest v3.2 release, it should support failover. But, "client talking to gRPC Proxy" may not work as well as regular client talking to core etcd server, because we do not support keepalive ping in gRPC proxy layer yet. Need more tests around gRPC proxy.

**Pull request merged**